text/0193-removal-of-type-unions.md

```diff
@@ -0,0 +1,575 @@
+```
+
+#### Recommended Approach
+
+For those cases, one solution is to have users of statically typed languages use _[escape hatches]_ to forcefully
+override properties with the desired late-bound value, since _[escape hatches]_ offer `any`-typed accessors. This
+results in a degradation of ergonomics for statically typed languages, but not not imply a redution in available
+features offered by the AWS CDK:
+
+- Where **TypeScript** developers could use:
+  ```ts
+  let cfnBucket: CfnBucket;
+  let value: IResolvable;
+  // ...
+  bucket.lifecycleConfiguration = value as any;
+  ```
+
+- Developers in **Java** (and other statically typed languages) would instead use:
+  ```java
+  CfnBucket bucket;
+  IResolvable resolvable;
+  // ...
+  bucket.addPropertyOverride('LifecycleConfiguration', resolvable);
+  ```
+
+#### Alternative Approach
+
+Another option is to generate specific wrapper types that effectively encapsulate the _type union_, however such wrapper
+types must then be used for _all_ properties in order to avoid incurring an API breaking change if some _CloudFormation
+Resource_ schema adds support for alternate value types on a given property (this is possible since CloudFormation
+describes resource schemas using [JSON Schema]). This would result in a significant deterioration of the ergonomics of
+the _CloudFormation Resource_ classes, but provides an API that is safe to use from any programming language:
+
+- This involves defining a type such as:
+  ```ts
+  // The type could be nested under the `Cfn<Resource>` class:
+  namespace CfnBucket {
+    // TODO: Come up with a better naming strategy
+    export class LifecycleConfigurationPropertyValue implements IResolvable {
+      /**
+       * Wraps an object literal to register on the Cfn property.
+      */
+      public static fromImmediate(value: LifecycleConfigurationProperty) {
+        return new LifecycleConfigurationPropertyValue(() => value);
+      }
+
+      /**
+       * Wraps an IResolvable late-bound value to register on the Cfn property.
+      */
+      public static fromLazy(value: IResolvable) {
+        return new LifecycleConfigurationPropertyValue((ctx) => value.resolve(ctx));
+      }
+
+      private constructor(public readonly resolve: (context: IResolveContext) => any) {}
+    }
+  }
+  ```
+
+- But consequently forces **TypeScript** users to replace:
+  ```ts
+  new CfnBucket(this, 'Resource', {
+    // ...
+    lifecycleConfiguration: {
+      rules: [{ expirationInDays: 7 }],
+    },
+  });
+  ```
+
+- With the pretty heavy syntax:
+  ```ts
+  new CfnBucket(this, 'Resource', {
+    // ...
+    lifecycleConfiguration: LifecycleConfigurationPropertyValue.immediate({
+      rules: ArrayOfRulesPropertyValue.immediate([
+        RulePropertyValue.immediate({ expirationInDays: 7 }),
+      ]),
+    }),
+  });
+  ```
+
+[escape hatches]: https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html
+[json schema]: https://json-schema.org
+
```

Comment:
I have another alternative.

What if we forced union types in JSII-compatible TypeScript to go through a type alias, with a JSII-specific annotation? Something like:


```ts
@JsiiTypeUnion
type LifecycleConfigurationPropertyValue = LifecycleConfigurationProperty | IResolvable;
```

And the type of `lifecycleConfiguration` in `CfnBucketProps` will be `readonly lifecycleConfiguration: LifecycleConfigurationPropertyValue`. Using type unions without an alias would be a JSII error.

Then, for Java, we would generate the "union-like" pattern class like you showed above, but the TypeScript / JavaScript experience would be untouched.

Is this feasible?

Reply:
I think that's definitely worth exploring as an alternative

Reply:
A `@decorator` **cannot** be applied to entities that have no materialization in Javascript (the `@decorator` is applied at runtime, not compile time). But that isn't limiting this (instead of annotating it with a `@decorator` it should simply be `export`ed to achieve the desired intent -- providing a name for the union). This suggestion is already documented under *Alternatives Considered*:

https://github.com/aws/aws-cdk-rfcs/blob/54295bf96daf868967a235edd96333bc084d9ef8/text/0193-removal-of-type-unions.md#L431-L443

---

This definitely allows generating union-like types in target languages where this is necessary.

However, that only solves the code generation problem, but does not solve the kernel ambiguity (which is the cause of the currently observed bugs caused by the presence of type unions): how can we reliably offer inspection methods that can conclusively tell overlapping interface (or struct) instances apart, so as to allow users to author *safe* code.

This would need to be complemented by enhancements in the *jsii kernel* to support such types, understanding that it is not possible to distinguish homonym methods with different argument signatures (since those are not available in the javascript runtime) without introducing additional code (possibly via a TypeScript transform - but this is going to be quite an endeavor).